diff --git a/src/GeofieldMapFieldTrait.php b/src/GeofieldMapFieldTrait.php
index ee78952..68c030a 100644
--- a/src/GeofieldMapFieldTrait.php
+++ b/src/GeofieldMapFieldTrait.php
@@ -628,10 +628,14 @@ trait GeofieldMapFieldTrait {
    * Form element json format validation handler.
    */
   public static function jsonValidate($element, FormStateInterface &$form_state) {
+    $element_values_array = JSON::decode($element['#value']);
     // Check the jsonValue.
-    if (!empty($element['#value']) && JSON::decode($element['#value']) == NULL) {
+    if (!empty($element['#value']) && $element_values_array == NULL) {
       $form_state->setError($element, t('The @field field is not valid Json Format.', ['@field' => $element['#title']]));
     }
+    else {
+      $form_state->setValueForElement($element, JSON::encode($element_values_array));
+    }
   }
 
   /**
